UdfSchemaProvider doesn't work with Structs

**Describe the bug**

`@UdfSchemaProvider` should be usable with UDFs that return `Struct`, but it fails to register such UDFs.

**To Reproduce**

The following UDF fails to register:
```java
  @UdfDescription(
      name = "KsqlStructUdf",
      description = "A test-only UDF for testing struct return types")
  public static class KsqlStructUdf {

    private static final SqlStruct RETURN =
        SqlStruct.builder().field("VAL", SqlTypes.STRING).build();

    @UdfSchemaProvider
    public SqlType provide(final List<SqlType> params) {
      return RETURN;
    }

    @Udf(schemaProvider = "provide")
    public Struct getDecimalStruct() {
      return null;
    }
  }
```

**Expected behavior**
Should register

**Actual behaviour**
```
[2020-05-05 15:11:23,955] WARN Failed to add UDF to the MetaStore. name=KsqlStructUdf method=public org.apache.kafka.connect.data.Struct io.confluent.ksql.function.UdfLoaderTest$KsqlStructUdf.getDecimalStruct() (io.confluent.ksql.function.UdfLoader:113)
io.confluent.ksql.util.KsqlException: Could not load UDF method with signature: public org.apache.kafka.connect.data.Struct io.confluent.ksql.function.UdfLoaderTest$KsqlStructUdf.getDecimalStruct()
	at io.confluent.ksql.function.FunctionLoaderUtils.getReturnType(FunctionLoaderUtils.java:176)
	at io.confluent.ksql.function.FunctionLoaderUtils.getReturnType(FunctionLoaderUtils.java:161)
	at io.confluent.ksql.function.UdfLoader.createFunction(UdfLoader.java:148)
	at io.confluent.ksql.function.UdfLoader.loadUdfFromClass(UdfLoader.java:106)
	at io.confluent.ksql.function.UserFunctionLoader.loadFunctions(UserFunctionLoader.java:122)
	at io.confluent.ksql.function.UserFunctionLoader.load(UserFunctionLoader.java:84)
	at io.confluent.ksql.function.UdfLoaderTest.initializeFunctionRegistry(UdfLoaderTest.java:1293)
	at io.confluent.ksql.function.UdfLoaderTest.<clinit>(UdfLoaderTest.java:98)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.junit.runners.BlockJUnit4ClassRunner.createTest(BlockJUnit4ClassRunner.java:250)
	at org.junit.runners.BlockJUnit4ClassRunner.createTest(BlockJUnit4ClassRunner.java:260)
	at org.junit.runners.BlockJUnit4ClassRunner$2.runReflectiveCall(BlockJUnit4ClassRunner.java:309)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.BlockJUnit4ClassRunner.methodBlock(BlockJUnit4ClassRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
Caused by: io.confluent.ksql.util.KsqlException: Type inference is not supported for: class org.apache.kafka.connect.data.Struct
	at io.confluent.ksql.execution.function.UdfUtil.handleParameterizedType(UdfUtil.java:138)
	at io.confluent.ksql.execution.function.UdfUtil.getSchemaFromType(UdfUtil.java:100)
	at io.confluent.ksql.function.FunctionLoaderUtils.getReturnType(FunctionLoaderUtils.java:170)
	... 32 more
```

--- a/ksqldb-common/src/main/java/io/confluent/ksql/function/types/StructType.java
+++ b/ksqldb-common/src/main/java/io/confluent/ksql/function/types/StructType.java
@@ -22,6 +22,11 @@ import java.util.stream.Collectors;
 
 public final class StructType extends ObjectType {
 
+  /**
+   * An empty struct accepts any struct as an instantiation.
+   */
+  public static final StructType ANY_STRUCT = StructType.builder().build();
+
   private final ImmutableMap<String, ParamType> schema;
 
   private StructType(final Map<String, ParamType> schema) {

--- a/ksqldb-engine/src/test/java/io/confluent/ksql/function/UdfLoaderTest.java
+++ b/ksqldb-engine/src/test/java/io/confluent/ksql/function/UdfLoaderTest.java
@@ -257,6 +257,20 @@ public class UdfLoaderTest {
   }
 
   @Test
+  public void shouldLoadFunctionWithStructSchemaProvider() {
+    // Given:
+    final UdfFactory returnDecimal = FUNC_REG.getUdfFactory(FunctionName.of("KsqlStructUdf"));
+
+    // When:
+    final List<SqlType> args = ImmutableList.of();
+    final KsqlScalarFunction function = returnDecimal.getFunction(args);
+
+    // Then:
+    assertThat(function.getReturnType(args), equalTo(KsqlStructUdf.RETURN));
+  }
+
+
+  @Test
   public void shouldLoadFunctionWithNestedDecimalSchema() {
     // Given:
     final UdfFactory returnDecimal = FUNC_REG.getUdfFactory(FunctionName.of("decimalstruct"));
@@ -269,7 +283,6 @@ public class UdfLoaderTest {
         function.getReturnType(ImmutableList.of()),
         equalTo(SqlStruct.builder().field("VAL", SqlDecimal.of(64, 2)).build()));
   }
-
 
   @Test
   public void shouldThrowOnReturnTypeMismatch() {
@@ -1388,6 +1401,25 @@ public class UdfLoaderTest {
       Struct struct = new Struct(schema);
       struct.put("VAL", BigDecimal.valueOf(123.45).setScale(2, RoundingMode.CEILING));
       return struct;
+    }
+  }
+
+  @UdfDescription(
+      name = "KsqlStructUdf",
+      description = "A test-only UDF for testing struct return types")
+  public static class KsqlStructUdf {
+
+    private static final SqlStruct RETURN =
+        SqlStruct.builder().field("VAL", SqlTypes.STRING).build();
+
+    @UdfSchemaProvider
+    public SqlType provide(final List<SqlType> params) {
+      return RETURN;
+    }
+
+    @Udf(schemaProvider = "provide")
+    public Struct getDecimalStruct() {
+      return null;
     }
   }
 

--- a/ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/TestUdf.java
+++ b/ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/TestUdf.java
@@ -15,11 +15,19 @@
 
 package io.confluent.ksql.function.udf;
 
+import io.confluent.ksql.schema.ksql.types.SqlStruct;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.schema.ksql.types.SqlTypes;
+import java.util.List;
+import org.apache.kafka.connect.data.SchemaBuilder;
 import org.apache.kafka.connect.data.Struct;
 
 @UdfDescription(name="test_udf", description = "test")
 @SuppressWarnings("unused")
 public class TestUdf {
+
+  private static final SqlStruct RETURN =
+      SqlStruct.builder().field("A", SqlTypes.STRING).build();
 
   @Udf(description = "returns the method name")
   public String doStuffIntString(final int arg1, final String arg2) {
@@ -46,5 +54,17 @@ public class TestUdf {
       @UdfParameter(schema = "STRUCT<A VARCHAR>") final Struct struct
   ) {
     return struct.getString("A");
+  }
+
+  @Udf(description = "returns the value of 'STRUCT<A VARCHAR>'", schemaProvider = "structProvider")
+  public Struct returnStructStuff() {
+    return new Struct(
+        SchemaBuilder.struct().field("A", SchemaBuilder.OPTIONAL_STRING_SCHEMA).optional().build()
+    ).put("A", "foo");
+  }
+
+  @UdfSchemaProvider
+  public SqlType structProvider(final List<SqlType> params) {
+    return RETURN;
   }
 }

--- a/ksqldb-execution/src/main/java/io/confluent/ksql/execution/function/UdfUtil.java
+++ b/ksqldb-execution/src/main/java/io/confluent/ksql/execution/function/UdfUtil.java
@@ -23,6 +23,7 @@ import io.confluent.ksql.function.types.MapType;
 import io.confluent.ksql.function.types.ParamType;
 import io.confluent.ksql.function.types.ParamTypes;
 import io.confluent.ksql.function.types.StringType;
+import io.confluent.ksql.function.types.StructType;
 import io.confluent.ksql.name.FunctionName;
 import io.confluent.ksql.util.KsqlException;
 import java.lang.reflect.GenericArrayType;
@@ -32,6 +33,7 @@ import java.lang.reflect.TypeVariable;
 import java.math.BigDecimal;
 import java.util.List;
 import java.util.Map;
+import org.apache.kafka.connect.data.Struct;
 
 public final class UdfUtil {
 
@@ -104,7 +106,9 @@ public final class UdfUtil {
     return schema;
   }
 
+  // CHECKSTYLE_RULES.OFF: CyclomaticComplexity
   private static ParamType handleParameterizedType(final Type type) {
+    // CHECKSTYLE_RULES.ON: CyclomaticComplexity
     if (type instanceof ParameterizedType) {
       final ParameterizedType parameterizedType = (ParameterizedType) type;
       if (parameterizedType.getRawType() == Map.class) {
@@ -133,6 +137,11 @@ public final class UdfUtil {
       return ArrayType.of(
           GenericType.of(
               ((GenericArrayType) type).getGenericComponentType().getTypeName()));
+    } else if (type instanceof Class<?> && Struct.class.isAssignableFrom((Class<?>) type)) {
+      // we don't have enough information here to return a more specific type of struct,
+      // but there are other parts of the code that enforce having a schema provider or
+      // schema annotation if a struct is being used
+      return StructType.ANY_STRUCT;
     }
 
     throw new KsqlException("Type inference is not supported for: " + type);

--- a/ksqldb-execution/src/test/java/io/confluent/ksql/execution/function/UdfUtilTest.java
+++ b/ksqldb-execution/src/test/java/io/confluent/ksql/execution/function/UdfUtilTest.java
@@ -25,12 +25,14 @@ import io.confluent.ksql.function.types.GenericType;
 import io.confluent.ksql.function.types.MapType;
 import io.confluent.ksql.function.types.ParamType;
 import io.confluent.ksql.function.types.ParamTypes;
+import io.confluent.ksql.function.types.StructType;
 import io.confluent.ksql.name.FunctionName;
 import io.confluent.ksql.util.KsqlException;
 import java.lang.reflect.Type;
 import java.math.BigDecimal;
 import java.util.List;
 import java.util.Map;
+import org.apache.kafka.connect.data.Struct;
 import org.hamcrest.CoreMatchers;
 import org.hamcrest.MatcherAssert;
 import org.junit.Test;
@@ -170,6 +172,14 @@ public class UdfUtilTest {
     assertThat(
         UdfUtil.getSchemaFromType(String.class),
         equalTo(ParamTypes.STRING)
+    );
+  }
+
+  @Test
+  public void shouldGetStringSchemaFromStructClass() {
+    assertThat(
+        UdfUtil.getSchemaFromType(Struct.class),
+        equalTo(StructType.ANY_STRUCT)
     );
   }
 

--- a/ksqldb-functional-tests/src/test/resources/historical_plans/struct-udfs_-_Output_struct/6.0.0_1588882528086/plan.json
+++ b/ksqldb-functional-tests/src/test/resources/historical_plans/struct-udfs_-_Output_struct/6.0.0_1588882528086/plan.json
@@ -1,0 +1,125 @@
+{
+  "plan" : [ {
+    "@type" : "ksqlPlanV1",
+    "statementText" : "CREATE STREAM TEST (VAL STRING) WITH (KAFKA_TOPIC='test_topic', VALUE_FORMAT='JSON');",
+    "ddlCommand" : {
+      "@type" : "createStreamV1",
+      "sourceName" : "TEST",
+      "schema" : "`ROWKEY` STRING KEY, `VAL` STRING",
+      "topicName" : "test_topic",
+      "formats" : {
+        "keyFormat" : {
+          "format" : "KAFKA"
+        },
+        "valueFormat" : {
+          "format" : "JSON"
+        }
+      }
+    }
+  }, {
+    "@type" : "ksqlPlanV1",
+    "statementText" : "CREATE STREAM OUTPUT AS SELECT TEST_UDF() VALUE\nFROM TEST TEST\nEMIT CHANGES",
+    "ddlCommand" : {
+      "@type" : "createStreamV1",
+      "sourceName" : "OUTPUT",
+      "schema" : "`ROWKEY` STRING KEY, `VALUE` STRUCT<`A` STRING>",
+      "topicName" : "OUTPUT",
+      "formats" : {
+        "keyFormat" : {
+          "format" : "KAFKA"
+        },
+        "valueFormat" : {
+          "format" : "JSON"
+        }
+      }
+    },
+    "queryPlan" : {
+      "sources" : [ "TEST" ],
+      "sink" : "OUTPUT",
+      "physicalPlan" : {
+        "@type" : "streamSinkV1",
+        "properties" : {
+          "queryContext" : "OUTPUT"
+        },
+        "source" : {
+          "@type" : "streamSelectV1",
+          "properties" : {
+            "queryContext" : "Project"
+          },
+          "source" : {
+            "@type" : "streamSourceV1",
+            "properties" : {
+              "queryContext" : "KsqlTopic/Source"
+            },
+            "topicName" : "test_topic",
+            "formats" : {
+              "keyFormat" : {
+                "format" : "KAFKA"
+              },
+              "valueFormat" : {
+                "format" : "JSON"
+              }
+            },
+            "sourceSchema" : "`ROWKEY` STRING KEY, `VAL` STRING"
+          },
+          "selectExpressions" : [ "TEST_UDF() AS VALUE" ]
+        },
+        "formats" : {
+          "keyFormat" : {
+            "format" : "KAFKA"
+          },
+          "valueFormat" : {
+            "format" : "JSON"
+          }
+        },
+        "topicName" : "OUTPUT"
+      },
+      "queryId" : "CSAS_OUTPUT_0"
+    }
+  } ],
+  "configs" : {
+    "ksql.extension.dir" : "ext",
+    "ksql.streams.cache.max.bytes.buffering" : "0",
+    "ksql.security.extension.class" : null,
+    "ksql.transient.prefix" : "transient_",
+    "ksql.persistence.wrap.single.values" : "true",
+    "ksql.authorization.cache.expiry.time.secs" : "30",
+    "ksql.schema.registry.url" : "",
+    "ksql.streams.default.deserialization.exception.handler" : "io.confluent.ksql.errors.LogMetricAndContinueExceptionHandler",
+    "ksql.output.topic.name.prefix" : "",
+    "ksql.streams.auto.offset.reset" : "earliest",
+    "ksql.query.pull.enable.standby.reads" : "false",
+    "ksql.connect.url" : "http://localhost:8083",
+    "ksql.service.id" : "some.ksql.service.id",
+    "ksql.internal.topic.min.insync.replicas" : "1",
+    "ksql.streams.shutdown.timeout.ms" : "300000",
+    "ksql.internal.topic.replicas" : "1",
+    "ksql.insert.into.values.enabled" : "true",
+    "ksql.query.pull.max.allowed.offset.lag" : "9223372036854775807",
+    "ksql.query.pull.max.qps" : "2147483647",
+    "ksql.streams.default.production.exception.handler" : "io.confluent.ksql.errors.ProductionExceptionHandlerUtil$LogAndFailProductionExceptionHandler",
+    "ksql.access.validator.enable" : "auto",
+    "ksql.streams.bootstrap.servers" : "localhost:0",
+    "ksql.streams.commit.interval.ms" : "2000",
+    "ksql.metric.reporters" : "",
+    "ksql.query.pull.metrics.enabled" : "false",
+    "ksql.streams.auto.commit.interval.ms" : "0",
+    "ksql.metrics.extension" : null,
+    "ksql.streams.topology.optimization" : "all",
+    "ksql.hidden.topics" : "_confluent.*,__confluent.*,_schemas,__consumer_offsets,__transaction_state,connect-configs,connect-offsets,connect-status,connect-statuses",
+    "ksql.streams.num.stream.threads" : "4",
+    "ksql.timestamp.throw.on.invalid" : "false",
+    "ksql.authorization.cache.max.entries" : "10000",
+    "ksql.metrics.tags.custom" : "",
+    "ksql.pull.queries.enable" : "true",
+    "ksql.udfs.enabled" : "true",
+    "ksql.udf.enable.security.manager" : "true",
+    "ksql.connect.worker.config" : "",
+    "ksql.any.key.name.enabled" : "false",
+    "ksql.sink.window.change.log.additional.retention" : "1000000",
+    "ksql.readonly.topics" : "_confluent.*,__confluent.*,_schemas,__consumer_offsets,__transaction_state,connect-configs,connect-offsets,connect-status,connect-statuses",
+    "ksql.udf.collect.metrics" : "false",
+    "ksql.persistent.prefix" : "query_",
+    "ksql.query.persistent.active.limit" : "2147483647"
+  }
+}

--- a/ksqldb-functional-tests/src/test/resources/historical_plans/struct-udfs_-_Output_struct/6.0.0_1588882528086/spec.json
+++ b/ksqldb-functional-tests/src/test/resources/historical_plans/struct-udfs_-_Output_struct/6.0.0_1588882528086/spec.json
@@ -1,0 +1,67 @@
+{
+  "version" : "6.0.0",
+  "timestamp" : 1588882528086,
+  "path" : "query-validation-tests/struct-udfs.json",
+  "schemas" : {
+    "CSAS_OUTPUT_0.KsqlTopic.Source" : "STRUCT<VAL VARCHAR> NOT NULL",
+    "CSAS_OUTPUT_0.OUTPUT" : "STRUCT<VALUE STRUCT<A VARCHAR>> NOT NULL"
+  },
+  "testCase" : {
+    "name" : "Output struct",
+    "inputs" : [ {
+      "topic" : "test_topic",
+      "key" : "1",
+      "value" : {
+        "val" : "foo"
+      },
+      "timestamp" : 0
+    } ],
+    "outputs" : [ {
+      "topic" : "OUTPUT",
+      "key" : "1",
+      "value" : {
+        "VALUE" : {
+          "A" : "foo"
+        }
+      },
+      "timestamp" : 0
+    } ],
+    "topics" : [ {
+      "name" : "OUTPUT",
+      "replicas" : 1,
+      "numPartitions" : 4
+    }, {
+      "name" : "test_topic",
+      "replicas" : 1,
+      "numPartitions" : 4
+    } ],
+    "statements" : [ "CREATE STREAM test (val VARCHAR) WITH (kafka_topic='test_topic', value_format='JSON');", "CREATE STREAM OUTPUT AS SELECT TEST_UDF() AS VALUE FROM test;" ],
+    "post" : {
+      "topics" : {
+        "topics" : [ {
+          "name" : "OUTPUT",
+          "keyFormat" : {
+            "formatInfo" : {
+              "format" : "KAFKA"
+            }
+          },
+          "valueFormat" : {
+            "format" : "JSON"
+          },
+          "partitions" : 4
+        }, {
+          "name" : "test_topic",
+          "keyFormat" : {
+            "formatInfo" : {
+              "format" : "KAFKA"
+            }
+          },
+          "valueFormat" : {
+            "format" : "JSON"
+          },
+          "partitions" : 4
+        } ]
+      }
+    }
+  }
+}

--- a/ksqldb-functional-tests/src/test/resources/historical_plans/struct-udfs_-_Output_struct/6.0.0_1588882528086/topology
+++ b/ksqldb-functional-tests/src/test/resources/historical_plans/struct-udfs_-_Output_struct/6.0.0_1588882528086/topology
@@ -1,0 +1,13 @@
+Topologies:
+   Sub-topology: 0
+    Source: KSTREAM-SOURCE-0000000000 (topics: [test_topic])
+      --> KSTREAM-TRANSFORMVALUES-0000000001
+    Processor: KSTREAM-TRANSFORMVALUES-0000000001 (stores: [])
+      --> Project
+      <-- KSTREAM-SOURCE-0000000000
+    Processor: Project (stores: [])
+      --> KSTREAM-SINK-0000000003
+      <-- KSTREAM-TRANSFORMVALUES-0000000001
+    Sink: KSTREAM-SINK-0000000003 (topic: OUTPUT)
+      <-- Project
+

--- a/ksqldb-functional-tests/src/test/resources/query-validation-tests/struct-udfs.json
+++ b/ksqldb-functional-tests/src/test/resources/query-validation-tests/struct-udfs.json
@@ -28,6 +28,19 @@
       "outputs": [
         {"topic":  "OUTPUT", "key": "1", "value": {"VALUE":  "1"}, "timestamp":  0}
       ]
+    },
+    {
+      "name": "Output struct",
+      "statements": [
+        "CREATE STREAM test (val VARCHAR) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT TEST_UDF() AS VALUE FROM test;"
+      ],
+      "inputs": [
+        {"topic":  "test_topic", "key": "1", "value": {"val": "foo"}, "timestamp":  0}
+      ],
+      "outputs": [
+        {"topic":  "OUTPUT", "key": "1", "value": {"VALUE": {"A": "foo"}}, "timestamp":  0}
+      ]
     }
   ]
 }

--- a/ksqldb-udf/src/main/java/io/confluent/ksql/function/udf/UdfSchemaProvider.java
+++ b/ksqldb-udf/src/main/java/io/confluent/ksql/function/udf/UdfSchemaProvider.java
@@ -20,14 +20,14 @@ import java.lang.annotation.Retention;
 import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
 
-@Retention(RetentionPolicy.RUNTIME)
-@Target({ElementType.METHOD})
 /**
  * The {@code UdfSchemaProvider} annotation on a method tells KSQL to use this method to resolve
  * the return type of the udf at runtime.
  *
  * <p>The corresponding udf annotation must have the {@code schemaProvider} specified.
  */
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD})
 public @interface UdfSchemaProvider {
 
 }

